repos/system_upgrade/common/actors/targetuserspacecreator/libraries/userspacegen.py

```diff
@@ -328,68 +332,187 @@ def _get_files_owned_by_rpms(context, dirpath, pkgs=None, recursive=False):
+        # end.
+        canonical_pointee_dir, pointee_filename = os.path.split(pointee_as_abspath)
+        canonical_pointee_dir = os.path.realpath(canonical_pointee_dir)
+
+        if canonical_pointee_dir.startswith(canonical_srcdir):
+            # Absolute path inside of the correct dir so we need to link to it
+            # But we need to determine what the link path should be before
+            # returning.
+
+            # Construct a relative path that points from the symlinks directory
+            # to the pointee.
+            link_to = os.readlink(symlink)
+            canonical_symlink_dir = os.path.realpath(os.path.dirname(symlink))
+            relative_path = os.path.relpath(canonical_pointee_dir, canonical_symlink_dir)
+
+            if link_to.startswith('/'):
+                # The original symlink was an absolute path so we will set this
+                # one to absolute too
+                # Note: Because absolute paths are constructed inside of
+                # srcdir, the relative path that we need to join here has to be
+                # relative to srcdir, not the directory that the symlink is
+                # being created in.
+                relative_to_srcdir = os.path.relpath(canonical_pointee_dir, canonical_srcdir)
+                corrected_path = os.path.normpath(os.path.join(srcdir, relative_to_srcdir, pointee_filename))
+
+            else:
+                # If the original link is a relative link, then we want the new
+                # link to be relative as well
+                corrected_path = os.path.normpath(os.path.join(relative_path, pointee_filename))
+
+            return ("link", corrected_path)
+
+        # pointee is a symlink that points outside of the srcdir so continue to
+        # the next symlink in the chain.
+
+    # The file is not a link so copy it
+    return ('copy', pointee_as_abspath)
+
+
 def _copy_decouple(srcdir, dstdir):
     """
-    Copy `srcdir` to `dstdir` while decoupling symlinks.
+    Copy files inside of `srcdir` to `dstdir` while decoupling symlinks.
 
     What we mean by decoupling the `srcdir` is that any symlinks pointing
     outside the directory will be copied as regular files. This means that the
     directory will become independent from its surroundings with respect to
     symlinks. Any symlink (or symlink chains) within the directory will be
     preserved.
 
+    .. warning::
+        `dstdir` must already exist.
     """
+    symlinks_to_process = []
+    for root, directories, files in os.walk(srcdir):
+        # relative path from srcdir because srcdir is replaced with dstdir for
+        # the copy.
+        relpath = os.path.relpath(root, srcdir)
+
+        # Create all directories with proper permissions for security
+        # reasons (Putting private data into directories that haven't had their
+        # permissions set appropriately may leak the private information.)
```

Comment:
It's good to thing about that. It's possible that such a function (or part of it) could be used for other things in future (cannot predict) and it's good to stay secure :muscle:  In case of /var/lib/leapp we do not have to be so strict, as the content inside of this dir could contain potentially sensitive data:
```
# ls -lZd /var/lib/leapp/
drwx------. root root system_u:object_r:var_lib_t:s0   /var/lib/leapp/
```

so the content inside should be secured.